File: Suyeon/ch14_Tree/1038_Binary_Search_Tree_to_Greater_Sum_Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    sum = 0
    def bstToGst(self, root: TreeNode) -> TreeNode:

        def inorder_traversal(root):
            if root is None:
                return

            inorder_traversal(root.right)
            self.sum += root.val
            root.val = self.sum
            inorder_traversal(root.left)

        inorder_traversal(root)

        return root

File: Sujin/Ch14_BinaryTree&BinarySearchTree/1038_Binary_Search_Tree_to_Greater_Sum_Tree.cpp
class Solution {
public:
    int sum=0;
    
    void sum_DFS(TreeNode* root){
        if(!root) return;
        
        if(root) sum += root->val;
        
        sum_DFS(root->left);
        sum_DFS(root->right);
    }
    
    void change_DFS(TreeNode* root){ // 전위순회로 트리 바꾸기
        if(!root) return;
        
        change_DFS(root->left);
        int tmp = root->val;
        root->val = sum;
        sum -= tmp;
        change_DFS(root->right);
        
    }
    
    TreeNode* bstToGst(TreeNode* root) {
        sum_DFS(root);
        change_DFS(root);
        return root;
    }
};